### PR TITLE
Update colshape and marker hit detection when attaching

### LINKED ISSUE #1317: Collision positions for colshapes are not updated when attached to an element
When attaching a colshape to an object, the collision position for the colshape is not updated, so onColShapeHit will not trigger when walking into the "visual" position of the colshape (viewed with /showcol).

I've included a test resource below. I'm running `v1.5.7-release-20358.5.001`

[colshape.zip](https://github.com/multitheftauto/mtasa-blue/files/4378360/colshape.zip)

### PR DESCRIPTION
**Pull request by Lpsd**
Fixes #1317

### CODE REVIEW

**Comment by qaisjp** [comment]
does this fix markers too?

**Comment by Lpsd** [comment]
I haven't tested with markers - it's getting pretty late so I'll investigate that tomorrow.

**Comment by Lpsd** [comment]
I've implemented this for markers too (it had the same issue as colshapes).

**Review by qaisjp — APPROVED** [review]
lgtm

**Comment by qaisjp** [comment]
have you tried overloading `AttachTo`? it's being called 2 lines above your additions

```bash
qaisjp@EdinburPro: ~/Documents/mtasa-blue git:(master) ✗
➜ ag "void AttachTo"
Server/mods/deathmatch/logic/CBlip.h
33:    void AttachTo(CElement* pElement);

Client/mods/deathmatch/logic/CClientPickup.h
109:    void AttachTo(CClientEntity* pEntity) override;

Client/mods/deathmatch/logic/CClientObject.h
51:    void AttachTo(CClientEntity* pEntity) override;




qaisjp@EdinburPro: ~/Documents/mtasa-blue git:(master) ✗
➜ ag "::AttachTo"
Server/mods/deathmatch/logic/CBlip.cpp
162:void CBlip::AttachTo(CElement* pElement)
164:    CElement::AttachTo(pElement);

Server/mods/deathmatch/logic/CElement.cpp
1080:void CElement::AttachTo(CElement* pElement)

Client/mods/deathmatch/logic/CClientPickup.cpp
95:void CClientPickup::AttachTo(CClientEntity* pEntity)
97:    CClientEntity::AttachTo(pEntity);

Client/mods/deathmatch/logic/CClientCamera.cpp
292:void CClientCamera::AttachTo(CElement* pElement)
301:    CClientEntity::AttachTo(pElement);

Client/mods/deathmatch/logic/CClientObject.cpp
177:void CClientObject::AttachTo(CClientEntity* pEntity)
179:    CClientEntity::AttachTo(pEntity);

Client/mods/deathmatch/logic/CClientEntity.cpp
605:void CClientEntity::AttachTo(CClientEntity* pEntity)
```

**Comment by Lpsd** [comment]
I've changed the implementation as suggested.

File: Client/mods/deathmatch/logic/CClientMarker.cpp [hunk 1]
```diff
@@ -89,6 +89,18 @@ void CClientMarker::SetPosition(const CVector& vecPosition)
     UpdateStreamPosition(vecPosition);
 }
 
+void CClientMarker::AttachTo(CClientEntity* pEntity)
+{
+    CClientEntity::AttachTo(pEntity);
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientMarker.cpp]: Have you considered just calling `m_pCollision->AttachTo(pEntity)` here?

File: Client/mods/deathmatch/logic/CClientMarker.cpp [hunk 2]
```diff
@@ -89,6 +89,18 @@ void CClientMarker::SetPosition(const CVector& vecPosition)
...
+    if (m_pAttachedToEntity)
+    {
+        CVector pVector;
+
+        pEntity->GetPosition(pVector);
+        SetPosition(pVector + m_vecAttachedPosition);
+    }
```
Comment (qaisjp) [inline Client/mods/deathmatch/logic/CClientMarker.cpp]:
```suggestion
    m_pCollision->AttachTo(pEntity);
```
Have you considered just calling `m_pCollision->AttachTo(pEntity)`?

**Review by ghost — APPROVED** [review]
Nice work! @qaisjp looks good. Should we merge it?

**Comment by Lpsd** [comment]
@saml1er there's a problem with a crash upon exiting MTA which seems to have been introduced by these changes. I'm still investigating -  see below:

```
Version = 1.5.7-custom.0.000
Time = Sat Mar 28 01:59:43 2020
Module = C:\Users\Ryan\Documents\GitHub\mtasa-blue\Bin\mods\deathmatch\client.dll
Code = 0xC0000005
Offset = 0x0004CCE5

EAX=00000000  EBX=1A3B6901  ECX=00000000  EDX=00000000  ESI=2680EB58
EDI=2669A3C0  EBP=0177F280  ESP=0177F280  EIP=7BCACCE5  FLG=00210246
CS=0023   DS=002B  SS=002B  ES=002B   FS=0053  GS=002B
```

File: Server/mods/deathmatch/logic/CColShape.cpp [hunk 3]
```diff
@@ -57,6 +57,16 @@ void CColShape::SetPosition(const CVector& vecPosition)
     CStaticFunctionDefinitions::RefreshColShapeColliders(this);
 }
 
+void CColShape::AttachTo(CElement* pElement)
+{
+    CElement::AttachTo(pElement);
+    if (m_pAttachedTo)
```
Comment (Lpsd) [inline Server/mods/deathmatch/logic/CColShape.cpp]: I believe the following change will resolve the crash I mentioned (the same in CClientColShape.cpp)
```cpp
if (pElement && m_pAttachedTo)
```

The reason being is that the destructor passes `NULL` to the `AttachTo` method, so we'd be trying to call `GetPosition` on `NULL`. I'll confirm this later on when I'm home.

**Comment by Lpsd** [comment]
I've fixed the crash mentioned above with the latest commit. Ready to go.

**Review by qaisjp — APPROVED** [review]
:shipit:

**Pull request merged**